File: vlibras_player/ios/Classes/VLibrasPlayerPlugin.swift
import Flutter
import UIKit
import WebKit

/// Flutter plugin that wraps VLibras via a WKWebView (iOS side).
public class VLibrasPlayerPlugin: NSObject, FlutterPlugin {

    // MARK: - Registration

    public static func register(with registrar: FlutterPluginRegistrar) {
        let methodChannel = FlutterMethodChannel(
            name: "vlibras/methods",
            binaryMessenger: registrar.messenger()
        )
        let eventChannel = FlutterEventChannel(
            name: "vlibras/events",
            binaryMessenger: registrar.messenger()
        )

        let instance = VLibrasPlayerPlugin()
        registrar.addMethodCallDelegate(instance, channel: methodChannel)
        eventChannel.setStreamHandler(instance)
    }

    // MARK: - Private state

    private var webView: WKWebView?
    private var eventSink: FlutterEventSink?

    // MARK: - FlutterMethodCallDelegate

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = call.arguments as? [String: Any]

        switch call.method {
        case "initialize":
            let avatar   = args?["avatar"]   as? String  ?? "icaro"
            let speed    = args?["speed"]    as? Double  ?? 1.0
            let autoPlay = args?["autoPlay"] as? Bool    ?? false
            let baseUrl  = args?["baseUrl"]  as? String  ?? "https://vlibras.gov.br/app"
            initialize(baseUrl: baseUrl, avatar: avatar, speed: speed, autoPlay: autoPlay, result: result)

        case "translate":
            let text = args?["text"] as? String ?? ""
            translate(text: text, result: result)

        case "show":
            setVisible(true)
            result(nil)

        case "hide":
            setVisible(false)
            result(nil)

        case "dispose":
            destroyWebView()
            result(nil)

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - Private helpers

    private func initialize(
        baseUrl: String,
        avatar: String,
        speed: Double,
        autoPlay: Bool,
        result: @escaping FlutterResult
    ) {
        destroyWebView()

        let config = WKWebViewConfiguration()
        config.allowsInlineMediaPlayback = true
        config.mediaTypesRequiringUserActionForPlayback = []

        let contentController = WKUserContentController()
        // Receive messages from JavaScript
        contentController.add(self, name: "vlibrasReady")
        contentController.add(self, name: "vlibrasTranslateComplete")
        contentController.add(self, name: "vlibrasError")
        config.userContentController = contentController

        let wv = WKWebView(frame: .zero, configuration: config)
        wv.isHidden = false
        wv.navigationDelegate = self
        wv.scrollView.isScrollEnabled = false
        self.webView = wv

        let html = buildHostHtml(baseUrl: baseUrl, avatar: avatar, speed: speed, autoPlay: autoPlay)
        wv.loadHTMLString(html, baseURL: URL(string: "https://vlibras.gov.br"))

        result(nil)
    }

    private func translate(text: String, result: @escaping FlutterResult) {
        guard let wv = webView else {
            result(FlutterError(code: "NOT_INITIALIZED",
                                message: "Call initialize() first",
                                details: nil))
            return
        }
        let escaped = text
            .replacingOccurrences(of: "'", with: "\\'")
            .replacingOccurrences(of: "\n", with: "\\n")
        let js = "if(window.__vlibrasPlayer) window.__vlibrasPlayer.translate('\(escaped)');"
        wv.evaluateJavaScript(js, completionHandler: nil)
        result(nil)
    }

    private func setVisible(_ visible: Bool) {
        webView?.isHidden = !visible
        emitEvent(type: visible ? "shown" : "hidden")
    }

    private func destroyWebView() {
        webView?.stopLoading()
        webView?.navigationDelegate = nil
        webView?.configuration.userContentController.removeAllScriptMessageHandlers()
        webView = nil
    }

    private func emitEvent(type: String, message: String? = nil) {
        var payload: [String: Any] = ["type": type]
        if let msg = message { payload["message"] = msg }
        DispatchQueue.main.async { [weak self] in
            self?.eventSink?(payload)
        }
    }

    private func buildHostHtml(
        baseUrl: String,
        avatar: String,
        speed: Double,
        autoPlay: Bool
    ) -> String {
        return """
        <!DOCTYPE html>
        <html lang="pt-BR">
        <head>
          <meta charset="UTF-8"/>
          <meta name="viewport" content="width=device-width,initial-scale=1"/>
          <style>
            html, body { margin:0; padding:0; background:transparent; overflow:hidden; }
            [vw] { position:fixed; bottom:0; right:0; width:100%; height:100%; }
            [vw] .active { display:flex!important; }
          </style>
        </head>
        <body>
          <div vw class="enabled">
            <div vw-access-button class="active"></div>
            <div vw-plugin-wrapper>
              <div class="vw-plugin-top-wrapper"></div>
            </div>
          </div>
          <script>
            window.__vlibrasConfig = { avatar: '\(avatar)', speed: \(speed), autoPlay: \(autoPlay) };

            (function() {
              var s = document.createElement('script');
              s.src = '\(baseUrl)/vlibras-plugin.js';
              s.onload = function() {
                new window.VLibras.Widget('\(baseUrl)');
                window.webkit.messageHandlers.vlibrasReady.postMessage({});
              };
              s.onerror = function() {
                window.webkit.messageHandlers.vlibrasError
                  .postMessage({ message: 'Failed to load vlibras-plugin.js' });
              };
              document.head.appendChild(s);
            })();
          </script>
        </body>
        </html>
        """
    }
}

// MARK: - WKScriptMessageHandler

extension VLibrasPlayerPlugin: WKScriptMessageHandler {
    public func userContentController(
        _ userContentController: WKUserContentController,
        didReceive message: WKScriptMessage
    ) {
        switch message.name {
        case "vlibrasReady":
            emitEvent(type: "ready")
        case "vlibrasTranslateComplete":
            emitEvent(type: "translateComplete")
        case "vlibrasError":
            let body = message.body as? [String: Any]
            emitEvent(type: "error", message: body?["message"] as? String ?? "Unknown error")
        default:
            break
        }
    }
}

// MARK: - WKNavigationDelegate

extension VLibrasPlayerPlugin: WKNavigationDelegate {}

// MARK: - FlutterStreamHandler

extension VLibrasPlayerPlugin: FlutterStreamHandler {
    public func onListen(
        withArguments arguments: Any?,
        eventSink events: @escaping FlutterEventSink
    ) -> FlutterError? {
        self.eventSink = events
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        self.eventSink = nil
        return nil
    }
}

File: vlibras_player/android/src/main/kotlin/br/gov/vlibras/vlibras_player/VLibrasPlayerPlugin.kt
package br.gov.vlibras.vlibras_player

import android.content.Context
import android.webkit.JavascriptInterface
import android.webkit.JsPromptResult
import android.webkit.JsResult
import android.webkit.WebChromeClient
import android.webkit.WebSettings
import android.webkit.WebView
import android.webkit.WebViewClient
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

/** Flutter plugin that wraps VLibras via a headless WebView. */
class VLibrasPlayerPlugin : FlutterPlugin, MethodCallHandler {

    private lateinit var methodChannel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private var eventSink: EventChannel.EventSink? = null

    private var webView: WebView? = null
    private var context: Context? = null

    // ── FlutterPlugin ──────────────────────────────────────────────────────────

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        context = binding.applicationContext

        methodChannel = MethodChannel(binding.binaryMessenger, "vlibras/methods")
        methodChannel.setMethodCallHandler(this)

        eventChannel = EventChannel(binding.binaryMessenger, "vlibras/events")
        eventChannel.setStreamHandler(object : EventChannel.StreamHandler {
            override fun onListen(arguments: Any?, sink: EventChannel.EventSink?) {
                eventSink = sink
            }
            override fun onCancel(arguments: Any?) {
                eventSink = null
            }
        })
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        destroyWebView()
    }

    // ── MethodCallHandler ──────────────────────────────────────────────────────

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "initialize" -> {
                val avatar  = call.argument<String>("avatar")  ?: "icaro"
                val speed   = call.argument<Double>("speed")   ?: 1.0
                val autoPlay = call.argument<Boolean>("autoPlay") ?: false
                val baseUrl = call.argument<String>("baseUrl") ?: "https://vlibras.gov.br/app"
                initialize(baseUrl, avatar, speed, autoPlay, result)
            }
            "translate" -> {
                val text = call.argument<String>("text") ?: ""
                translate(text, result)
            }
            "show"    -> { setVisible(true);  result.success(null) }
            "hide"    -> { setVisible(false); result.success(null) }
            "dispose" -> { destroyWebView();  result.success(null) }
            else      -> result.notImplemented()
        }
    }

    // ── Private helpers ────────────────────────────────────────────────────────

    private fun initialize(
        baseUrl: String,
        avatar: String,
        speed: Double,
        autoPlay: Boolean,
        result: Result,
    ) {
        destroyWebView()
        val ctx = context ?: run { result.error("NO_CONTEXT", "Context not available", null); return }

        val wv = WebView(ctx).also { webView = it }

        wv.settings.apply {
            javaScriptEnabled = true
            domStorageEnabled = true
            mediaPlaybackRequiresUserGesture = false
            cacheMode = WebSettings.LOAD_DEFAULT
            // NEVER_ALLOW (não ALWAYS_ALLOW): a página é servida via HTTPS
            // (vlibras.gov.br); não há motivo para permitir sub-recursos HTTP
            // misturados, e ALWAYS_ALLOW é uma superfície de risco desnecessária.
            mixedContentMode = WebSettings.MIXED_CONTENT_NEVER_ALLOW
        }
        // WebChromeClient customizado: a classe padrão (sem overrides) mostra
        // diálogos NATIVOS do Android para window.alert()/confirm()/prompt() —
        // texto de JS cru, por cima de todo o app, bloqueando a thread principal
        // até o usuário tocar OK. Este WebView é headless/embutido: nenhum
        // diálogo de JS deve escapar para a UI do sistema.
        wv.webChromeClient = object : WebChromeClient() {
            override fun onJsAlert(
                view: WebView?,
                url: String?,
                message: String?,
                result: JsResult?,
            ): Boolean {
                result?.confirm()
                return true
            }

            override fun onJsConfirm(
                view: WebView?,
                url: String?,
                message: String?,
                result: JsResult?,
            ): Boolean {
                result?.confirm()
                return true
            }

            override fun onJsPrompt(
                view: WebView?,
                url: String?,
                message: String?,
                defaultValue: String?,
                result: JsPromptResult?,
            ): Boolean {
                result?.cancel()
                return true
            }
        }
        wv.addJavascriptInterface(
            VLibrasJsInterface(
                onReady           = { emitEvent("ready") },
                onTranslateComplete = { emitEvent("translateComplete") },
                onError           = { msg -> emitEvent("error", message = msg) },
            ),
            "VLibrasFlutter",
        )

        wv.webViewClient = object : WebViewClient() {
            override fun onPageFinished(view: WebView?, url: String?) {
                // Inject VLibras plugin script and bootstrap
                val js = """
                    (function() {
                        var script = document.createElement('script');
                        script.src = '$baseUrl/vlibras-plugin.js';
                        script.onload = function() {
                            new window.VLibras.Widget('$baseUrl');
                            var player = document.querySelector('[vw] .active');
                            VLibrasFlutter.onReady();
                        };
                        script.onerror = function(e) {
                            VLibrasFlutter.onError('Failed to load vlibras-plugin.js');
                        };
                        document.head.appendChild(script);
                    })();
                """.trimIndent()
                view?.evaluateJavascript(js, null)
            }
        }

        // Load a minimal HTML host page
        wv.loadDataWithBaseURL(
            "https://vlibras.gov.br",
            buildHostHtml(avatar, speed, autoPlay),
            "text/html",
            "utf-8",
            null,
        )

        result.success(null)
    }

    private fun translate(text: String, result: Result) {
        val wv = webView ?: run {
            result.error("NOT_INITIALIZED", "Call initialize() first", null)
            return
        }
        val escaped = text.replace("'", "\\'").replace("\n", "\\n")
        wv.evaluateJavascript(
            "if(window.__vlibrasPlayer) window.__vlibrasPlayer.translate('$escaped');",
            null,
        )
        result.success(null)
    }

    private fun setVisible(visible: Boolean) {
        webView?.visibility = if (visible)
            android.view.View.VISIBLE
        else
            android.view.View.INVISIBLE
        emitEvent(if (visible) "shown" else "hidden")
    }

    private fun destroyWebView() {
        webView?.destroy()
        webView = null
    }

    private fun emitEvent(type: String, message: String? = null) {
        val payload = mutableMapOf<String, Any?>("type" to type)
        if (message != null) payload["message"] = message
        // Must be called on the main thread
        android.os.Handler(android.os.Looper.getMainLooper()).post {
            eventSink?.success(payload)
        }
    }

    private fun buildHostHtml(avatar: String, speed: Double, autoPlay: Boolean): String = """
        <!DOCTYPE html>
        <html lang="pt-BR">
        <head>
          <meta charset="UTF-8"/>
          <meta name="viewport" content="width=device-width,initial-scale=1"/>
          <style>
            html, body { margin:0; padding:0; background:transparent; overflow:hidden; }
            [vw] { position:fixed; bottom:0; right:0; width:100%; height:100%; }
            [vw] .active { display:flex!important; }
          </style>
        </head>
        <body>
          <div vw class="enabled">
            <div vw-access-button class="active"></div>
            <div vw-plugin-wrapper>
              <div class="vw-plugin-top-wrapper"></div>
            </div>
          </div>
          <script>
            window.__vlibrasConfig = {
              avatar: '$avatar',
              speed: $speed,
              autoPlay: $autoPlay,
            };
          </script>
        </body>
        </html>
    """.trimIndent()
}

/** JavaScript → Kotlin bridge. All methods are called from the WebView JS thread. */
class VLibrasJsInterface(
    private val onReady: () -> Unit,
    private val onTranslateComplete: () -> Unit,
    private val onError: (String) -> Unit,
) {
    @JavascriptInterface
    fun onReady() = onReady.invoke()

    @JavascriptInterface
    fun onTranslateComplete() = onTranslateComplete.invoke()

    @JavascriptInterface
    fun onError(message: String) = onError.invoke(message)
}
